Add an Upload feature to program the keyboard from Easy AVR itself

mohitg11 opened the pull request:
-- Added a Upload menu which can be used to program the keyboard directly from Easy AVR.
-- Added option to build firmware as .bin file.
-- Packaged corresponding external tools required for programming in "exttools" directly
-- Added Help menu item for new Upload feature

Currently upload works perfectly, with following bootloaders:
- Teensy, LUFA HID - Using Teensy CLI mod
- Atmel DFU, LUFA DFU - Using dfu-programmer
- LUFA Mass Storage - Using OS copy

Only other commonly used bootloader not supported as of now, but planned:
- LUFA CDC (Used in ProMicro/Arduino/etc) - Uses serial communication

dhowland commented:
I have concerns about this pull request.

The biggies:

1) I'm not sure it's appropriate to be including these other tools in the code base.  I have not looked into the licenses of each of the tools but if they permit redistribution I would think they would require accompanied documentation for license terms and/or attribution.  Also, it is hardly ever a good idea to have executables in the source code.  The .hex files are already a stretch, but I do it to simplify distribution.  That brings up another point.  We don't want to have to support dfu-programmer when it isn't working right.  Finally, this is a cross-platform tool.  These .exes won't work on Linux or Mac.

2) None of the new features are used by any of the supported boards.  What is the need?

The rest:

3) There's a lot of code duplication.  I realize the tool suffers a little from this already, but there's no point in making it worse.
4) Spelling errors
5) The choice between build firmware and build binary will confuse users.  If .bin support was added to the tool, I would simply look at the user's chosen filename to decide which kind of file to make.  
6) It's been a while since I studied the docs, but I'm pretty sure subprocess is being used incorrectly.

mohitg11 commented:
> 1) I'm not sure it's appropriate to be including these other tools in the code base. I have not looked into the licenses of each of the tools but if they permit redistribution I would think they would require accompanied documentation for license terms and/or attribution. Also, it is hardly ever a good idea to have executables in the source code. The .hex files are already a stretch, but I do it to simplify distribution. That brings up another point. We don't want to have to support dfu-programmer when it isn't working right. Finally, this is a cross-platform tool. These .exes won't work on Linux or Mac.

All the tools are distributed under the GPL, which your application is distributed under as well, thus I don't think this should be an issue. dfu-programmer is working right, I don't understand what you mean when you say, "it isn't working right". Regarding the executables in the source code, surely we can choose to leave them out from the source code but have them in the windows package that you release with each update? I understand that the tool is cross-platform, but the releases you do, are always for Windows. This is entirely up to you. All the tools can be compiled for other OS's as well if the user wants them.

> 2) None of the new features are used by any of the supported boards. What is the need?

Upload feature can be used with all the supported boards.

> 3) There's a lot of code duplication. I realize the tool suffers a little from this already, but there's no point in making it worse.

I understand your concern, but don't believe it to be strong enough a reason to dismiss the added functionality. We can always do some code review later on.

> 4) Spelling errors

Is this for the readme? Surely this can be corrected.

> 5) The choice between build firmware and build binary will confuse users. If .bin support was added to the tool, I would simply look at the user's chosen filename to decide which kind of file to make.

Agreed, I too think it would be better to just have one menu item and the user can choose the file type when building the firmware.

> 6) It's been a while since I studied the docs, but I'm pretty sure subprocess is being used incorrectly.

Don't understand what you mean by this. I've run the code on three machines at home and it works fine on all of them. I did go through the documentation before using this as I'm programming in Python for the first time.

Finally, I really do think adding the Upload functionality would be a great addition and make the tool even Easier to use.

dhowland commented:
Let's take it one piece at a time.

Check out my new commit to master.  It adds .bin support in a way that is more consistent with convention and it only takes 11 lines changed, with no external utility.  It also makes overlay() more generic, which is a good thing.

mohitg11 commented:
I've refactored the code to be more optimized and removed code duplication as much as possible. 

I've now used your method for generation of the binary file.

I've corrected the spelling errors from the help file. 

I've modified the subprocess call so that it can be used on different platforms as the two tools used (HID_Bootloader and DFU_Programmer) can both be compiled for multiple OSes.

I've also excluded the executables from the source. I've included the source of the hid_bootloader from LUFA. I've not included the dfu-programmer code as it is a bigger package, but can be included (https://github.com/dfu-programmer/dfu-programmer) if you so wish. With respect to the Windows package you release, you can include the two executables in your package.

dhowland commented:
Cool, taking a look.  I was actually planning on working this some more but I got sick.  I was thinking for the executables the appropriate thing to do would be to include them in the windows build if the dev had downloaded them, but otherwise to just search the path.

mohitg11 commented:
Cool, do have a look and let me know of further changes.

I'll push a commit later tonight or tomorrow so that we run the external tools by using a modified environment, so if user has the tools added to system path, it can be called from there, or the exttools folder.

mohitg11 commented:
Just pushed another update that will allow uploading over serial port. The ProMicro which is used quite often for handwire boards as it is a cheaper alternative to the Teensy, uses this method.

dhowland commented:
EasyAVR is starting to attract users who don't know how to computer.  Going to have to get serious about this.  We have to make sure it's really solid, though.

Are you sure that the ProMicro requires serial port programming?  That doesn't sound right.  Doesn't it use the Arduino bootloader?

mohitg11 commented:
Haha, it was intended to be Easy to use ;) 

Positive about the ProMicro, it uses Caterina, which can be programmed over usb via serial com using avrdude emulating avr109 programmer.

dhowland commented:
Hey you should add attribution to the .c/.h/.py files to which you've made significant changes.  You don't have to if you don't want, though

mohitg11 commented:
I tested out the following over the weekend:
- Program Teensy with Teensy Bootloader (BL)
- Program Teensy++ with DFU BL
- Program Teensy++ with MSD BL
- Program Teensy++ with CDC BL
- Program Pro Micro with CDC BL

All worked without a hitch. The Pro Micro is finicky with resets without a physical switch as you have to short RST pin and GND twice quickly to get it to go in BL mode for 8 seconds. This issues occurs when programming it manually using avrdude or Arduino (Which also uses avrdude)

Regarding the attribution, it's fine, thanks! I'll add it in a future commit (I'm working on adding RGB LEDs right now)

mohitg11 commented:
Hey D,

Any updates on the code review?

dhowland commented:
Yeah, I've been trying to do this in between other projects.  I'm thinking that once these changes are in, we need to go bigger.  Basically what I'm talking about is almost a plugin system.

I think you probably wouldn't be surprised to learn, based on what I've said in the support thread, that I view programming as out of scope for the keymapper.  Still, a "save and program" function makes perfect sense and would be useful.

However, I think the users should be able to add on any additional function to the build, whether it be auto-programming, archiving/uploading to the cloud, or even additional modifications to the build.  Anything.  So there should be a general purpose interface that looks for add-ons and puts them in the new menu.  So if a user wanted to modify the programming flow, all they'd have to do is make their own script and throw it in the correct directory.

It requires some thought, though.  I've been pondering it for a while now.  I see two ways forward.  One is that all external functions are written in shell scripts.  This is definitely easiest for the user, however it will require all built-in scripts to be written in both sh and bat.  The other is to use python and have the tool load them the way it does the user boards.  Lots to think about.

mohitg11 commented:
Wow, that is definitely an interesting line of thought. It would require quite a bit of effort on your end to add that sort of modularity with the plugins and all. 

I would definitely go with the Python approach. To me it makes much more sense as the user can read the configurations from the python file directly, use the current GUI design as well as to certain degree deter novices from creating subpar plugins. 

Another issue I see, with the growing list of supported boards, you might need a new way of displaying the options :laughing: Just out of curiosity why do you have the "supported" boards listed in the **init**.py file? Why not just read the boards directly from the boards folder like you have it for new custom boards. THis would save the confusion of where to place custom files and the need to update the **init**.py file with each added board. Just feel it would be more easier for people to add custom boards this way. This reminds me, why don't you publish your releases via GitHub, I feel it will be better, if there is too much traffic on your DropBox, they disable your public links, has happened with me because of the GH Resources folder on my DropBox.

//Enough OT

dhowland commented:
They're in the init.py file because they are part of the package.  By using the Python import system to load the built-ins, I don't have to play any games trying to find where it's installed.  The user configs are intentionally in a different spot so that they are accessible from any version of the tool, even if there are multiple installations.

dhowland commented:
Sat down to work this today but apparently I don't know how to use Git.  I'm gonna go read a book or something...

mohitg11 commented:
Lol, what happened?

dhowland commented:
I am closing this pull request because it is obsoleted by the new programming framework.  I have coded Teensy loader and Flip for windows.  I have not completely reimplemented all of the changes from this pull request.  I encourage you to extend the new code.  What you have to do is make a new class that inherits from ProgrammingTask.  See the programming.py for examples.

mohitg11 commented:
Thanks, w8ll test it out later today. Excited!

mohitg11 commented:
From a very preliminary test, I can see one place of improvement. In the programming window, for selection of  programming method, right now if it is a Teensy device, it shows only that method as available, and FLIP/DFU for windows. I would recommend having all programming options available, with a default selections based on,
- If Teensy based device, Teensy is selected
- Else If Windows FLIP is selected
- Else DFU is selected

Reason is the hid_bootloader_cli.exe is compatible with LUFA HID bootloader and it's not a far fetched tough to think people might have changed the bootloader on their Teensy device (I have for one of my Teensy's). This would allow flexibility for future as well as keeping it noob-friendly by selecting the default programming method.

From the current methods implemented, only avrdude is missing, which I'll work on adding. One question, why do you have it setup for non-Windows devices. avrdude can be used on Windows, in fact it is with the Pro-Micros and Arduino.

dhowland commented:
You're right.  There's no reason to restrict user options unnecessarily.  I committed a fix for that.

AVRdude is the way it is because I don't really know anything about AVRdude.

mohitg11 commented:
Small correction to be done with your release package, the Teensy CLI exe is only usable with Teensy bootloader, not LUFA HID, you'll need the hid_bootloader_cli.exe for that. It is compatible with both.

Also, why don't you publish your releases via GitHub, I feel it will be better, if there is too much traffic on your DropBox, they disable your public links, has happened with me because of the GH Resources folder on my DropBox.
